In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# Data Visualization Libraries
import seaborn as sns
import matplotlib.pyplot as plt

# Check for Outliers and Skewness(0 > log transform; 0.5 > square root transform)
from scipy.stats import zscore, boxcox

In [2]:
#Import dataset
df=pd.read_csv('data.csv')
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0     1900.0        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0     1900.0        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
23673     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23674       NaN           NaN        NaN          NaN -37.87882   144.60184   
23675       NaN           NaN        NaN          NaN -37.72006   145.10547   
23676     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23677       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
23673   Western Metropolitan        5262.0  
23674   Western Metropolitan        5262.0  
23675  Northern Metropolitan        1369.0  
23676   Western Metropolitan        6543.0  
23677   Western Metropolitan        6543.0  

[23678 rows x 21 columns]

### Ans1. Determine the no. of rows and no. of columns in the dataset

In [3]:
df.shape

(23678, 21)

### Ans2. Display the datatypes of each column.

In [4]:
df.dtypes

Suburb            object
Address           object
Rooms            float64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

### Ans3. Determine the row indexes of rows having all null values (i.e. empty rows). 

In [5]:
d= df.loc[df.isnull().sum(axis=1)>20]

In [6]:
d

Suburb Address  Rooms Type  Price Method SellerG Date  Distance  \
31       NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
1145     NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
1348     NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
2119     NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
3277     NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
...      ...     ...    ...  ...    ...    ...     ...  ...       ...   
22288    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
22704    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
22918    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
23153    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   
23341    NaN     NaN    NaN  NaN    NaN    NaN     NaN  NaN       NaN   

       Postcode  ...  Bathroom  Car  Landsize  BuildingArea  YearBuilt  \
31          NaN  ...       NaN  NaN       NaN           NaN        NaN   
1145        NaN  ...       NaN  NaN       NaN           NaN        NaN   
1348        NaN  ...       NaN  NaN       NaN           NaN        NaN   
2119        NaN  ...       NaN  NaN       NaN           NaN        NaN   
3277        NaN  ...       NaN  NaN       NaN           NaN        NaN   
...         ...  ...       ...  ...       ...           ...        ...   
22288       NaN  ...       NaN  NaN       NaN           NaN        NaN   
22704       NaN  ...       NaN  NaN       NaN           NaN        NaN   
22918       NaN  ...       NaN  NaN       NaN           NaN        NaN   
23153       NaN  ...       NaN  NaN       NaN           NaN        NaN   
23341       NaN  ...       NaN  NaN       NaN           NaN        NaN   

       CouncilArea Lattitude  Longtitude  Regionname Propertycount  
31             NaN       NaN         NaN         NaN           NaN  
1145           NaN       NaN         NaN         NaN           NaN  
1348           NaN       NaN         NaN         NaN           NaN  
2119           NaN       NaN         NaN         NaN           NaN  
3277           NaN       NaN         NaN         NaN           NaN  
...            ...       ...         ...         ...           ...  
22288          NaN       NaN         NaN         NaN           NaN  
22704          NaN       NaN         NaN         NaN           NaN  
22918          NaN       NaN         NaN         NaN           NaN  
23153          NaN       NaN         NaN         NaN           NaN  
23341          NaN       NaN         NaN         NaN           NaN  

[131 rows x 21 columns]

In [7]:
d.shape

(131, 21)

Finding:-
    131 rows having all null value.

### Ans4. Determine the total number of empty rows and their percentage in whole dataset and remove them

In [8]:
df1=df.copy()

In [9]:
df1.shape

(23678, 21)

In [10]:
df1.dropna(how='all').shape

(23547, 21)

In [11]:
df1.shape

(23678, 21)

In [12]:
23678-23547

131

In [13]:
(131/23678)*100

0.5532561871779711

0.55% empty rows in the datasets.

In [14]:
df1.dropna(how='all',inplace=True)

In [15]:
df1.shape

(23547, 21)

### Ans5. Determine the column names having missing values.

In [16]:
df1.columns[df1.isnull().any()]

Index(['Price', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

### Ans6. Determine the missing values and their percentage in each column.

In [17]:
df1.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [18]:
(df1.isnull().sum()/len(df1))*100

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.875398
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.004247
Postcode          0.004247
Bedroom2         19.030025
Bathroom         19.042766
Car              19.645815
Landsize         26.062768
BuildingArea     57.455302
YearBuilt        50.991634
CouncilArea      33.511700
Lattitude        18.278337
Longtitude       18.278337
Regionname        0.004247
Propertycount     0.004247
dtype: float64

### Ans7. Determine the row indexes having missing values.

In [19]:
Row_Missing= df1.loc[df1.isnull().sum(axis=1)>0]

In [20]:
Row_Missing

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
5        Abbotsford    40 Federation La    3.0    h   850000.0     PI   
7        Abbotsford        16 Maugie St    4.0    h        NaN     SN   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0  ...       1.0  1.0   
3             Rounds  04-02-2016       2.5    3067.0  ...       2.0  1.0   
5             Biggin  04-03-2017       2.5    3067.0  ...       2.0  1.0   
7             Nelson  06-08-2016       2.5    3067.0  ...       2.0  2.0   
...              ...         ...       ...       ...  ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
5          94.0           NaN        NaN        Yarra -37.79690   144.99690   
7         400.0         220.0     2006.0        Yarra -37.79650   144.99650   
...         ...           ...        ...          ...       ...         ...   
23673     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23674       NaN           NaN        NaN          NaN -37.87882   144.60184   
23675       NaN           NaN        NaN          NaN -37.72006   145.10547   
23676     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23677       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
5      Northern Metropolitan        4019.0  
7      Northern Metropolitan        4019.0  
...                      ...           ...  
23673   Western Metropolitan        5262.0  
23674   Western Metropolitan        5262.0  
23675  Northern Metropolitan        1369.0  
23676   Western Metropolitan        6543.0  
23677   Western Metropolitan        6543.0  

[17351 rows x 21 columns]

Finding:-
    17351 rows having missing value.

### Ans8. Determine the missing values and their percentage in each row.

In [21]:
df1.isnull()

Suburb  Address  Rooms   Type  Price  Method  SellerG   Date  Distance  \
0       False    False  False  False   True   False    False  False     False   
1       False    False  False  False  False   False    False  False     False   
2       False    False  False  False  False   False    False  False     False   
3       False    False  False  False   True   False    False  False     False   
4       False    False  False  False  False   False    False  False     False   
...       ...      ...    ...    ...    ...     ...      ...    ...       ...   
23673   False    False  False  False   True   False    False  False     False   
23674   False    False  False  False   True   False    False  False     False   
23675   False    False  False  False  False   False    False  False     False   
23676   False    False  False  False  False   False    False  False     False   
23677   False    False  False  False  False   False    False  False     False   

       Postcode  ...  Bathroom    Car  Landsize  BuildingArea  YearBuilt  \
0         False  ...     False  False     False          True       True   
1         False  ...     False  False     False          True       True   
2         False  ...     False  False     False         False      False   
3         False  ...     False  False     False          True       True   
4         False  ...     False  False     False         False      False   
...         ...  ...       ...    ...       ...           ...        ...   
23673     False  ...     False  False     False         False      False   
23674     False  ...      True   True      True          True       True   
23675     False  ...     False  False      True          True       True   
23676     False  ...     False  False     False         False      False   
23677     False  ...     False  False      True         False      False   

       CouncilArea  Lattitude  Longtitude  Regionname  Propertycount  
0            False      False       False       False          False  
1            False      False       False       False          False  
2            False      False       False       False          False  
3            False      False       False       False          False  
4            False      False       False       False          False  
...            ...        ...         ...         ...            ...  
23673         True      False       False       False          False  
23674         True      False       False       False          False  
23675         True      False       False       False          False  
23676         True      False       False       False          False  
23677         True      False       False       False          False  

[23547 rows x 21 columns]

### Ans9. Determine the column names having more than 30% missing values and remove them from the dataset.

In [22]:
(df1.isnull().sum()/len(df1))*100>30

Suburb           False
Address          False
Rooms            False
Type             False
Price            False
Method           False
SellerG          False
Date             False
Distance         False
Postcode         False
Bedroom2         False
Bathroom         False
Car              False
Landsize         False
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude        False
Longtitude       False
Regionname       False
Propertycount    False
dtype: bool

In [23]:
df1.drop(columns=['BuildingArea','YearBuilt','CouncilArea'], axis=1, inplace=True)

In [24]:
df1.shape

(23547, 18)

### Ans10. Determine the number and percentage of rows having more than 5 missing values. If the percentage of such rows is less than 20% then remove all these rows.

In [25]:
row5_miss= df1.loc[df1.isnull().sum(axis=1)>5]

In [26]:
row5_miss

Suburb              Address  Rooms Type      Price Method  \
15            Abbotsford     217 Langridge St    3.0    h  1000000.0      S   
16            Abbotsford      18a Mollison St    2.0    t   745000.0      S   
19            Abbotsford  403/609 Victoria St    2.0    u   542000.0      S   
21            Abbotsford    25/84 Trenerry Cr    2.0    u   760000.0     SP   
22            Abbotsford    106/119 Turner St    1.0    u   481000.0     SP   
...                  ...                  ...    ...  ...        ...    ...   
23626  Templestowe Lower  1/207 Manningham Rd    2.0    u   550000.0     PI   
23630          Thornbury      1/128 Dundas St    2.0    t   770000.0     PI   
23633          Thornbury        111 Pender St    2.0    t   858000.0      S   
23639             Toorak   21/1059 Malvern Rd    2.0    u   720000.0     VB   
23662           Werribee         91 Latham St    4.0    h   540000.0     PI   

            SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
15           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
16           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
19           Dingle  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
21           Biggin  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
22     Purplebricks  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
...             ...         ...       ...       ...       ...       ...  ...   
23626         Barry  26-08-2017      12.4    3107.0       NaN       NaN  NaN   
23630       McGrath  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23633        Jellis  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23639        Beller  26-08-2017       4.1    3142.0       NaN       NaN  NaN   
23662       Triwest  26-08-2017      14.7    3030.0       NaN       NaN  NaN   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
15          NaN        NaN         NaN  Northern Metropolitan         4019.0  
16          NaN        NaN         NaN  Northern Metropolitan         4019.0  
19          NaN        NaN         NaN  Northern Metropolitan         4019.0  
21          NaN        NaN         NaN  Northern Metropolitan         4019.0  
22          NaN        NaN         NaN  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23626       NaN        NaN         NaN   Eastern Metropolitan         5420.0  
23630       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23633       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23639       NaN        NaN         NaN  Southern Metropolitan         7217.0  
23662       NaN        NaN         NaN   Western Metropolitan        16166.0  

[4278 rows x 18 columns]

In [27]:
row5_miss.shape

(4278, 18)

In [28]:
(4278/23547)*100

18.16791948018856

Conclusion:
    1. Row index having more than 5 null value is 4278.
    2. Row having more than 5 null value persent is 18% and less than 20% so it drop.

In [29]:
df1.drop(df1[df1.isnull().sum(axis=1)>5].index,axis=0,inplace=True)

In [30]:
df1.shape

(19269, 18)

### Ans11. Display the description of data and percentage of missing values still left column wise.

In [31]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19269 entries, 0 to 23677
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         19269 non-null  object 
 1   Address        19269 non-null  object 
 2   Rooms          19269 non-null  float64
 3   Type           19269 non-null  object 
 4   Price          15086 non-null  float64
 5   Method         19269 non-null  object 
 6   SellerG        19269 non-null  object 
 7   Date           19269 non-null  object 
 8   Distance       19269 non-null  float64
 9   Postcode       19269 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  Lattitude      19243 non-null  float64
 15  Longtitude     19243 non-null  float64
 16  Regionname     19269 non-null  object 
 17  Propertycount  19269 non-null  float64
dtypes: flo

In [32]:
df1.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price            4183
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2          203
Bathroom          206
Car               348
Landsize         1859
Lattitude          26
Longtitude         26
Regionname          0
Propertycount       0
dtype: int64

In [33]:
(df1.isnull().sum()/len(df1))*100

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.708444
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.000000
Postcode          0.000000
Bedroom2          1.053506
Bathroom          1.069075
Car               1.806010
Landsize          9.647621
Lattitude         0.134932
Longtitude        0.134932
Regionname        0.000000
Propertycount     0.000000
dtype: float64

### Ans12. Considering the 'Price' column as target variable which is to be predicted, how will you treat the missing values? Give a brief explanation and treat the missing values in 'Price' column.

Observation:
    Price column is a target column and in this column more than 20% null value so not drop all the null value of Price  columns.
    because if drop all the null value of the price columns so drop more data it's not good way.

In [34]:
df1.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  19269.000000  1.508600e+04  19269.000000  19269.000000  19066.000000   
mean       2.973117  1.066695e+06     10.204365   3110.015777      2.951956   
std        0.978969  6.389331e+05      6.065786     93.730487      0.996032   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.400000e+05      6.100000   3049.000000      2.000000   
50%        3.000000  9.000000e+05      9.200000   3102.000000      3.000000   
75%        4.000000  1.320000e+06     13.000000   3150.000000      4.000000   
max       12.000000  9.000000e+06     48.100000   3977.000000     30.000000   

           Bathroom           Car       Landsize     Lattitude    Longtitude  \
count  19063.000000  18921.000000   17410.000000  19243.000000  19243.000000   
mean       1.570897      1.626235     551.783458    -37.812434    145.000287   
std        0.712684      0.974048    3544.288014      0.079926      0.106071   
min        0.000000      0.000000       0.000000    -38.184180    144.431620   
25%        1.000000      1.000000     181.000000    -37.859300    144.939300   
50%        1.000000      2.000000     448.000000    -37.809700    145.004300   
75%        2.000000      2.000000     656.000000    -37.759800    145.063100   
max       12.000000     26.000000  433014.000000    -37.407580    145.526350   

       Propertycount  
count   19269.000000  
mean     7490.249520  
std      4319.566673  
min       129.000000  
25%      4442.000000  
50%      6567.000000  
75%     10331.000000  
max     21650.000000

In [35]:
df1['Price'].min()

85000.0

In [36]:
df1['Price'].max()

9000000.0

In [37]:
df1[df1['Price'].isna()]

Suburb             Address  Rooms Type  Price Method  \
0        Abbotsford       68 Studley St    2.0    h    NaN     SS   
3        Abbotsford  18/659 Victoria St    3.0    u    NaN     VB   
7        Abbotsford        16 Maugie St    4.0    h    NaN     SN   
8        Abbotsford        53 Turner St    2.0    h    NaN      S   
9        Abbotsford        99 Turner St    2.0    h    NaN      S   
...             ...                 ...    ...  ...    ...    ...   
23660      Werribee         6 Feijoa Ct    5.0    h    NaN      W   
23661      Werribee      11 Gundaroo Sq    4.0    h    NaN     PI   
23664      Werribee            3 Sky Ct    3.0    h    NaN     PI   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u    NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h    NaN      S   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
7             Nelson  06-08-2016       2.5    3067.0       3.0       2.0  2.0   
8             Biggin  06-08-2016       2.5    3067.0       4.0       1.0  2.0   
9            Collins  06-08-2016       2.5    3067.0       3.0       2.0  1.0   
...              ...         ...       ...       ...       ...       ...  ...   
23660            YPA  26-08-2017      14.7    3030.0       5.0       3.0  2.0   
23661      Stockdale  26-08-2017      14.7    3030.0       4.0       2.0  6.0   
23664          Barry  26-08-2017      14.7    3030.0       3.0       2.0  2.0   
23673      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0       NaN       NaN  NaN   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
7         400.0  -37.79650   144.99650  Northern Metropolitan         4019.0  
8         201.0  -37.79950   144.99740  Northern Metropolitan         4019.0  
9         202.0  -37.79960   144.99890  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23660     535.0  -37.89362   144.66767   Western Metropolitan        16166.0  
23661       NaN  -37.91230   144.66980   Western Metropolitan        16166.0  
23664       NaN  -37.89096   144.64200   Western Metropolitan        16166.0  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23674       NaN  -37.87882   144.60184   Western Metropolitan         5262.0  

[4183 rows x 18 columns]

In [38]:
#price column missing value fillup by mean 
df1['Price'].fillna(df1['Price'].mean(),inplace=True)

In [39]:
df1.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2          203
Bathroom          206
Car               348
Landsize         1859
Lattitude          26
Longtitude         26
Regionname          0
Propertycount       0
dtype: int64

Price column null value is filling by mean beacuse median value is very large 

### Ans13. Now check the percentage of missing values in each column again, if missing values exist, treat it with the most suitable method.

In [40]:
df1.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2          203
Bathroom          206
Car               348
Landsize         1859
Lattitude          26
Longtitude         26
Regionname          0
Propertycount       0
dtype: int64

In [41]:
(df1.isnull().sum()/len(df1))*100

Suburb           0.000000
Address          0.000000
Rooms            0.000000
Type             0.000000
Price            0.000000
Method           0.000000
SellerG          0.000000
Date             0.000000
Distance         0.000000
Postcode         0.000000
Bedroom2         1.053506
Bathroom         1.069075
Car              1.806010
Landsize         9.647621
Lattitude        0.134932
Longtitude       0.134932
Regionname       0.000000
Propertycount    0.000000
dtype: float64

In [43]:
#traet the null value
df1.dtypes

Suburb            object
Address           object
Rooms            float64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

In [44]:
df1.columns[df1.isnull().any()]

Index(['Bedroom2', 'Bathroom', 'Car', 'Landsize', 'Lattitude', 'Longtitude'], dtype='object')

No null Value persent in Object type columns.

In [45]:
#ALL numerical columns missing value fillup by mean 
df1['Bedroom2'].fillna(df1['Bedroom2'].mean(),inplace=True)
df1['Bathroom'].fillna(df1['Bathroom'].mean(),inplace=True)

df1['Car'].fillna(df1['Car'].mean(),inplace=True)
df1['Landsize'].fillna(df1['Landsize'].mean(),inplace=True)

df1['Lattitude'].fillna(df1['Lattitude'].mean(),inplace=True)
df1['Longtitude'].fillna(df1['Longtitude'].mean(),inplace=True)

In [46]:
df1.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

### Ans14. Now since you have completed your data cleaning, determine the number and percentage of columns removed.

In [47]:
df.shape

(23678, 21)

In [48]:
df1.shape

(19269, 18)

Conclusion:
    3 columns removed after data cleaning.

In [49]:
(3/21)*100

14.285714285714285

### Ans15. Determine the number and percentage of rows removed in whole process. Also comment briefly that according to you the amount of data lost is justified or not.

In [50]:
df.shape

(23678, 21)

In [51]:
df1.shape

(19269, 18)

In [52]:
23678-19269

4409

4409 rows removed in whole process.

In [53]:
(4409/23678)*100

18.620660528760872

### 4409 rows and 18% of row removed in whole data cleaning process.data lost is more than 10%  its not good But in this dataset lots of null value persent and not drop any neccesary data. So not lossing information. According to me in this dataset less than 25% data lost and is justified.
